File: data/src/notFoss/java/org/cryptomator/data/cloud/googledrive/GoogleDriveCloudNodeFactory.java
package org.cryptomator.data.cloud.googledrive;

import com.google.api.services.drive.model.File;

import org.cryptomator.util.Optional;

import java.util.Date;

class GoogleDriveCloudNodeFactory {

	public static GoogleDriveFile file(GoogleDriveFolder parent, File file) {
		return new GoogleDriveFile(parent, file.getName(), getNodePath(parent, file.getName()), file.getId(), getFileSize(file), getModified(file));
	}

	public static GoogleDriveFile file(GoogleDriveFolder parent, String name, Optional<Long> size) {
		return new GoogleDriveFile(parent, name, getNodePath(parent, name), null, size, Optional.empty());
	}

	private static Optional<Date> getModified(File file) {
		return file.getModifiedTime() != null ? Optional.of(new Date(file.getModifiedTime().getValue())) : Optional.empty();
	}

	private static Optional<Long> getFileSize(File file) {
		return file.getSize() != null ? Optional.of(file.getSize()) : Optional.empty();
	}

	public static GoogleDriveFile file(GoogleDriveFolder parent, String name, Optional<Long> size, String path, String driveId) {
		return new GoogleDriveFile(parent, name, path, driveId, size, Optional.empty());
	}

	public static GoogleDriveFolder folder(GoogleDriveFolder parent, File file) {
		return new GoogleDriveFolder(parent, file.getName(), getNodePath(parent, file.getName()), file.getId());
	}

	public static GoogleDriveFolder folder(GoogleDriveFolder parent, String name) {
		return new GoogleDriveFolder(parent, name, getNodePath(parent, name), null);
	}

	public static GoogleDriveFolder folder(GoogleDriveFolder parent, String name, String path, String driveId) {
		return new GoogleDriveFolder(parent, name, path, driveId);
	}

	public static GoogleDriveNode from(GoogleDriveFolder parent, File file) {
		if (isFolder(file)) {
			return folder(parent, file);
		} else {
			return file(parent, file);
		}
	}

	public static boolean isFolder(File file) {
		return file.getMimeType().equals("application/vnd.google-apps.folder");
	}

	public static String getNodePath(GoogleDriveFolder parent, String name) {
		return parent.getPath() + "/" + name;
	}

}

File: data/src/apkStorePlaystore/java/org/cryptomator/data/cloud/googledrive/GoogleDriveCloudNodeFactory.kt
package org.cryptomator.data.cloud.googledrive

import com.google.api.services.drive.model.File
import java.util.Date

internal object GoogleDriveCloudNodeFactory {

	fun file(parent: GoogleDriveFolder, file: File): GoogleDriveFile {
		return GoogleDriveFile(parent, file.name, getNodePath(parent, file.name), file.id, getFileSize(file), getModified(file))
	}

	fun file(parent: GoogleDriveFolder, name: String, size: Long?): GoogleDriveFile {
		return GoogleDriveFile(parent, name, getNodePath(parent, name), null, size, null)
	}

	private fun getModified(file: File): Date? {
		return if (file.modifiedTime != null) Date(file.modifiedTime.value) else null
	}

	private fun getFileSize(file: File): Long? {
		return if (file.getSize() != null) file.getSize() else null
	}

	fun file(parent: GoogleDriveFolder, name: String, size: Long?, path: String, driveId: String): GoogleDriveFile {
		return GoogleDriveFile(parent, name, path, driveId, size, null)
	}

	fun folder(parent: GoogleDriveFolder, file: File): GoogleDriveFolder {
		return GoogleDriveFolder(parent, file.name, getNodePath(parent, file.name), file.id)
	}

	fun folder(parent: GoogleDriveFolder, name: String): GoogleDriveFolder {
		return GoogleDriveFolder(parent, name, getNodePath(parent, name), null)
	}

	fun folder(parent: GoogleDriveFolder?, name: String, path: String, driveId: String): GoogleDriveFolder {
		return GoogleDriveFolder(parent, name, path, driveId)
	}

	fun from(parent: GoogleDriveFolder, file: File): GoogleDriveNode {
		return when {
			isFolder(file) -> {
				folder(parent, file)
			}
			isShortcutFolder(file) -> {
				folder(parent, file.name, getNodePath(parent, file.name), file.shortcutDetails.targetId)
			}
			else -> {
				file(parent, file)
			}
		}
	}

	fun isFolder(file: File): Boolean {
		return file.mimeType == "application/vnd.google-apps.folder"
	}

	fun isShortcutFolder(file: File): Boolean {
		return file.mimeType == "application/vnd.google-apps.shortcut" && file.shortcutDetails.targetMimeType == "application/vnd.google-apps.folder"
	}

	fun getNodePath(parent: GoogleDriveFolder, name: String): String {
		return parent.path + "/" + name
	}
}
